# Behavioral Feature Analysis for Churn

#### The goal of this notebook is to identify behavioral patterns in customers that lead to churn.

### Approach

1. **Calculate churn rate** for each feature group.  
   Churn rate = proportion of customers who left within a group.  
   - Easy for categorical features.  
   - For numerical features, we first find thresholds where churn rates change significantly.

2. **Label groups with similar churn rates** to simplify interpretation.  
   - Example: Support Calls

| Support Calls | Churn Rate |
|---------------|------------|
| 1             | 0.450      |
| 2             | 0.452      |
| 3             | 0.80       |

   - Churn rate is stable for 1-2 calls but jumps sharply at 3 calls.  
   - Segmentation:  
     - 1-2 calls → Low Support Call  
     - 3 calls → High Support Call

3. **Three criteria to identify important features**  
   - The churn rate difference must be significant for at least one group compared to others.  
   - The pattern should be stable — not random noise.  
   - Group sizes should be comparable.

   - Example: Issue Level

| Issue Level | Churn Rate |
|-------------|------------|
| Low         | 0.1        |
| Medium      | 0.25       |
| High        | 0.6        |

   - Churn rate stays stable across low and medium but increases sharply at high issue level.

In [1]:
from sqlalchemy import create_engine

engine = create_engine(
    "mysql+pymysql://vivek:password123@localhost/churn_db"
)

In [2]:
import pymysql
import pandas as pd
conn = pymysql.connect(
    host="localhost",
    user="vivek",
    password="password123",
    database="churn_db"
)

In [3]:
query = """
SELECT * FROM churn LIMIT 10;
"""
df = pd.read_sql(query,engine)
df

,CustomerID,Age,Gender,Tenure,Usage_Frequency,Support_Calls,Payment_Delay,Subscription_Type,Contract_Length,Total_Spend,Last_Interaction,Churn
0,1,22,Female,25,14,4,27,Basic,Monthly,598.0,9,1.0
1,2,30,Female,39,14,5,18,Standard,Annual,932.0,17,1.0
2,2,41,Female,28,28,7,13,Standard,Monthly,584.0,20,0.0
3,3,47,Male,27,10,2,29,Premium,Annual,757.0,21,0.0
4,3,65,Female,49,1,10,8,Basic,Monthly,557.0,6,1.0
5,4,55,Female,14,4,6,18,Basic,Quarterly,185.0,3,1.0
6,4,35,Male,9,12,5,17,Premium,Quarterly,232.0,18,0.0
7,5,53,Female,58,24,9,2,Standard,Annual,533.0,18,0.0
8,5,58,Male,38,21,7,7,Standard,Monthly,396.0,29,1.0
9,6,23,Male,32,20,5,8,Basic,Monthly,617.0,20,1.0


In [4]:
query = """
SELECT Churn,COUNT(Churn) as total_customer FROM churn
GROUP BY Churn ;
"""
df = pd.read_sql(query,engine)
df

,Churn,total_customer
0,0.0,224714
1,1.0,280492


- Nearly 44.4% are churned customers, indicating the customer retention problem.

In [5]:
query = """
SELECT AVG(Age) as avg_age,AVG(Tenure) as avg_Ten,AVG(Usage_Frequency) as avg_Fre,AVG(Support_Calls) as avg_SC,AVG(Payment_Delay) as avg_PD,
AVG(Total_Spend) as avg_TS,AVG(Last_Interaction) as avg_LI FROM churn ;
"""
df = pd.read_sql(query,engine)
df

,avg_age,avg_Ten,avg_Fre,avg_SC,avg_PD,avg_TS,avg_LI
0,39.7042,31.3504,15.7148,3.8333,13.4968,620.072766,14.6106


In [6]:
query = """
SELECT 
    Churn,
    AVG(Tenure),
    AVG(Usage_Frequency),
    AVG(Support_Calls),
    AVG(Payment_Delay),
    AVG(Total_Spend),
    AVG(Last_Interaction)
FROM churn
GROUP BY Churn;
"""
df = pd.read_sql(query,engine)
df

,Churn,AVG(Tenure),AVG(Usage_Frequency),AVG(Support_Calls),AVG(Payment_Delay),AVG(Total_Spend),AVG(Last_Interaction)
0,0.0,31.7608,16.2277,2.0258,10.3830,721.394858,13.3877
1,1.0,31.0217,15.3039,5.2814,15.9915,538.899354,15.5903


#### Objective
- To identify the presence of behavioral differences between churned and non-churned customers for each feature.
#### Approach
1. The average value of each feature across churned and non-churned customers was calculated.
2. The percentage difference between churned and non-churned customers was computed.

##### Observations
- The percentage difference is highest for support calls with 89.1%, followed by payment delay with 42.53% and then next is total spend with 28.96%.
- The percentage differences for last interaction, usage frequency and tenure are relatively low with 15.2%, 5.86% and 2.35%.

 - Support calls, payment delay and total spend are more strongerly associated with customer churn than last interaction, tenure and usage frequency. 

In [7]:
query="""
SELECT 
    Support_Calls,
    AVG(CASE WHEN Churn = 1 THEN 1 ELSE 0 END) AS churn_rate
FROM churn
GROUP BY Support_Calls
ORDER BY Support_Calls;
"""
df = pd.read_sql(query,engine)
df

,Support_Calls,churn_rate
0,0,0.2988
1,1,0.2987
2,2,0.3098
3,3,0.4026
4,4,0.5537
5,5,0.8749
6,6,0.9134
7,7,0.9172
8,8,0.9141
9,9,0.9140


In [36]:
query="""
SELECT 
    COUNT(Support_Calls) AS number_of_customer
FROM churn
WHERE Support_Calls	BETWEEN 0 AND 2;
"""
df = pd.read_sql(query,engine)
df

,number_of_customer
0,220630


In [35]:
query="""
SELECT 
    COUNT(Support_Calls) AS number_of_customer
FROM churn
WHERE Support_Calls BETWEEN 3 AND 4;
"""
df = pd.read_sql(query,engine)
df

,number_of_customer
0,101350


In [37]:
query="""
SELECT 
    COUNT(Support_Calls) AS number_of_customer
FROM churn
WHERE Support_Calls BETWEEN	5 and 10;
"""
df = pd.read_sql(query,engine)
df

,number_of_customer
0,183226


#### Observations
1. Low level issue  (0-2): Churn rate remains constant at ~0.29, with ~220k customers.

2. Medium level issue (3-4): Churn rate rises from 0.40 to 0.55, with ~101k customers.

3. High level issue (5-10): Churn rate remains constant at ~0.91, with ~183k customers.



In [32]:
query="""
SELECT 
CASE 
    WHEN Support_Calls <= 2 THEN 'Low Issues'
    WHEN Support_Calls <= 4 THEN 'Medium Issues' 
    ELSE 'High Issues' 
END AS Issue_Level,
AVG(CASE WHEN Churn = 1 THEN 1 ELSE 0 END) as churn_rate,COUNT(*) as Number_of_customer
FROM churn
GROUP BY Issue_level;
"""
df = pd.read_sql(query,engine)
df

,Issue_Level,churn_rate,Number_of_customer
0,High Issues,0.9079,183226
1,Low Issues,0.3023,220630
2,Medium Issues,0.4681,101350


  #### Observations and Conclusion

1. The churn rate increases sharply from 0.46 at medium issue level to 0.90 high issue level.
    - This indicates that the customers at high issue level are likely to churn.

2. The percentage difference between low and medium issue levels is 43.09% but, it jumps to 63.94% between medium and high issue level.

3. The number of customers are comparable for each issue level.

In [9]:
query = """
SELECT Payment_Delay,AVG(CASE WHEN Churn=1 THEN 1 ELSE 0 END) AS churn_rate
FROM churn
GROUP BY Payment_Delay;
"""
df = pd.read_sql(query,engine)
df

,Payment_Delay,churn_rate
0,0,0.4347
1,1,0.4278
2,2,0.4376
3,3,0.4360
4,4,0.4323
5,5,0.4347
6,6,0.4353
7,7,0.4338
8,8,0.4397
9,9,0.4364


#### Observations
1. Low level delay(0-15): Churn rate remains constant at ~0.43, with ~296k customers.

2. Medium level delay (16-20): Churn rate remains constant at ~0.48, with ~96k customers.

3. High level issue (5-10): Churn rate remains constant at ~0.94, with ~111k customers.



In [10]:
query = """
SELECT 
CASE
    WHEN Payment_Delay <= 15 THEN "Low Delay"
    WHEN Payment_Delay <= 20 THEN "Medium Delay"
    ELSE "High Delay"
    END AS Delay_level,
AVG(CASE WHEN Churn = 1 THEN 1 ELSE 0 END) AS churn_rate,COUNT(*) as Number_of_customer
FROM churn
GROUP BY Delay_Level;
"""
df = pd.read_sql(query,engine)
df

,Delay_level,churn_rate,Number_of_customer
0,High Delay,0.9420,111709
1,Low Delay,0.4345,296988
2,Medium Delay,0.4791,96509


#### Observations and Conclusion

1. The churn rate increases sharply from 0.47 at the medium delay level to 0.94 at the high delay level.
    - This indicates that the customers at high delay level are likely to churn.

2. The percentage difference between low and medium delay levels is only 9.75% , but it jumps to 65.19% between medium and high delay levels.

3. The number of customers are comparable for for low and high delay levels , but slightly less for medium level.

In [11]:
query = """
SELECT Total_Spend,AVG(CASE WHEN Churn=1 THEN 1 ELSE 0 END) AS churn_rate
FROM churn
GROUP BY Total_Spend;
"""
df = pd.read_sql(query,engine)
df

,Total_Spend,churn_rate
0,100.00,0.8824
1,100.02,1.0000
2,100.06,1.0000
3,100.07,1.0000
4,100.08,1.0000
...,...,...
68358,999.96,0.1429
68359,999.97,0.0000
68360,999.98,0.0000
68361,999.99,0.0000


#### Observations
1. The total spend is continuous data.
2. It is difficult to identify groups in total spend due to 68363 rows, given by "GROUP BY Total_Spend".
    - This can be solved by computing deciles of sorted total spend.

In [ ]:
query = """
SELECT 
    spend_group,
    MIN(Total_Spend) AS min_spend,
    MAX(Total_Spend) AS max_spend,
    AVG(CASE WHEN Churn = 1 THEN 1 ELSE 0 END) AS churn_rate
FROM (
    SELECT 
        Total_Spend,
        Churn,
        NTILE(10) OVER (ORDER BY Total_Spend) AS spend_group
    FROM churn
) t
GROUP BY spend_group
ORDER BY spend_group;
"""
df = pd.read_sql(query,engine)
df

,spend_group,min_spend,max_spend,churn_rate
0,1,100.0,239.0,0.9030
1,2,239.0,378.0,0.8997
2,3,378.0,508.0,0.8471
3,4,508.0,578.0,0.4180
4,5,578.0,648.9,0.4123
5,6,648.9,719.0,0.4141
6,7,719.0,789.0,0.4182
7,8,789.0,859.0,0.4136
8,9,859.0,929.3,0.4120
9,10,929.3,1000.0,0.4140


#### Observations
1. Low level spend (100-508): Churn rate remains constant at ~0.88, with ~151k customers.

3. High level issue (508-1000): Churn rate remains constant at ~0.41, with ~353k customers.

In [13]:
query = """
SELECT 
CASE
    WHEN Total_Spend <= 508 THEN "Low Spend"
    ELSE "High Spend"
    END AS Spend_level,
AVG(CASE WHEN Churn = 1 THEN 1 ELSE 0 END) AS churn_rate,COUNT(*) as Number_of_customer
FROM churn
GROUP BY Spend_Level;
"""
df = pd.read_sql(query,engine)
df

,Spend_level,churn_rate,Number_of_customer
0,High Spend,0.4144,353515
1,Low Spend,0.8833,151691


#### Observations and Conclusion

1. The churn rate drops sharply from 0.88 at the low spend level to 0.41 at the high spend level.
    - This indicates that the customers at low spend level are likely to churn.

2. The percentage difference between low and high spend levels is a high 72%.

3. The number of customers are comparable for low and high spend levels.

In [14]:
query = """
SELECT Last_Interaction,AVG(CASE WHEN Churn=1 THEN 1 ELSE 0 END) AS churn_rate
FROM churn
GROUP BY Last_Interaction;
"""
df = pd.read_sql(query,engine)
df

,Last_Interaction,churn_rate
0,1,0.4871
1,2,0.4914
2,3,0.4947
3,4,0.4881
4,5,0.4907
5,6,0.4903
6,7,0.4927
7,8,0.4878
8,9,0.4946
9,10,0.4881


#### Observations
1. Low level spend (1-15): Churn rate remains constant at ~0.49, with ~282k customers.

3. High level issue (508-1000): Churn rate remains constant at ~0.63, with ~222k customers.

In [15]:
query = """
SELECT 
CASE
    WHEN Last_Interaction <= 15 THEN "Low LI"
    ELSE "High LI"
    END AS LI_level,
AVG(CASE WHEN Churn = 1 THEN 1 ELSE 0 END) AS churn_rate,COUNT(*) as Number_of_customer
FROM churn
GROUP BY LI_Level;
"""
df = pd.read_sql(query,engine)
df

,LI_level,churn_rate,Number_of_customer
0,High LI,0.6371,222600
1,Low LI,0.4907,282606


#### Observations and Conclusion

#####  LI - Last Interaction
1. The churn rate increases from 0.49 at the low LI level to 0.63 at the high LI level.

2. The percentage difference between low and high LI levels is only 25%.

3. The number of customers are comparable for low and high LI levels.

In [16]:
query = """
SELECT Tenure,AVG(CASE WHEN Churn=1 THEN 1 ELSE 0 END) AS churn_rate
FROM churn
GROUP BY Tenure;
"""
df = pd.read_sql(query,engine)
df

,Tenure,churn_rate
0,1,0.6166
1,2,0.6105
2,3,0.5948
3,4,0.6040
4,5,0.6069
5,6,0.5102
6,7,0.5191
7,8,0.5262
8,9,0.5092
9,10,0.5127


#### Observations - Tenure

1. **Very Low Tenure (1-5):** Churn rate ~0.60, with ~37k customers.
2. **Low Tenure (6-11):** Churn rate ~0.50, with ~51k customers.
3. **Medium Tenure (12-23):** Churn rate ~0.59, with ~98k customers.
4. **High Tenure (24-60):** Churn rate ~0.54, with ~317k customers.

**Conclusion:** The churn rate is relatively stable across tenure levels, with only a 10 percentage point range (0.50-0.60). This indicates tenure is not a primary driver of churn.

In [17]:
query = """
SELECT 
CASE
    WHEN Tenure <= 5 THEN " Very Low Tenure"
    WHEN Tenure <= 11 THEN "Low Tenure"
    WHEN Tenure <= 24 THEN"Medium Tenure"
    ELSE "High Tenure"
    END AS Tenure_Level,
AVG(CASE WHEN Churn = 1 THEN 1 ELSE 0 END) AS churn_rate,COUNT(*) as Number_of_customer
FROM churn
GROUP BY Tenure_Level;
"""
df = pd.read_sql(query,engine)
df

,Tenure_Level,churn_rate,Number_of_customer
0,Very Low Tenure,0.6065,37445
1,High Tenure,0.5441,317463
2,Low Tenure,0.5154,51494
3,Medium Tenure,0.5921,98804


- For every class inbetween there is not much difference in the churn rate so Tenure may not influencing that much customer churn

In [18]:
query = """
SELECT Usage_Frequency,AVG(CASE WHEN Churn=1 THEN 1 ELSE 0 END) AS churn_rate
FROM churn
GROUP BY Usage_Frequency;
"""
df = pd.read_sql(query,engine)
df

,Usage_Frequency,churn_rate
0,1,0.6295
1,2,0.6332
2,3,0.6082
3,4,0.6140
4,5,0.6130
5,6,0.5979
6,7,0.5767
7,8,0.5906
8,9,0.5890
9,10,0.5338


- The Usage Frequency 8 divides the churn rate in two classes.

In [19]:
query = """
SELECT 
CASE
    WHEN  Usage_Frequency <= 9 THEN "Low UF"
    ELSE "High UF"
    END AS UF_level,
AVG(CASE WHEN Churn = 1 THEN 1 ELSE 0 END) AS churn_rate,COUNT(*) as Number_of_customer
FROM churn
GROUP BY UF_Level;
"""
df = pd.read_sql(query,engine)
df

,UF_level,churn_rate,Number_of_customer
0,High UF,0.5349,361154
1,Low UF,0.6061,144052


In [20]:
query = """
SELECT Age,AVG(CASE WHEN Churn=1 THEN 1 ELSE 0 END) AS churn_rate
FROM churn
GROUP BY Age;
"""
df = pd.read_sql(query,engine)
df

,Age,churn_rate
0,18,0.6210
1,19,0.6171
2,20,0.5245
3,21,0.5279
4,22,0.5326
5,23,0.5319
6,24,0.5374
7,25,0.5268
8,26,0.5239
9,27,0.5265


In [21]:
query = """
SELECT 
CASE
    WHEN Age <= 19 THEN " Very Young"
    WHEN Age <= 29 THEN "Young Adult"
    WHEN Age <= 39 THEN "Old Adult"
    ELSE "Old"
    END AS Age_group,
AVG(CASE WHEN Churn = 1 THEN 1 ELSE 0 END) AS churn_rate,COUNT(*) as Number_of_customer
FROM churn
GROUP BY Age_group;
"""
df = pd.read_sql(query,engine)
df

,Age_group,churn_rate,Number_of_customer
0,Very Young,0.6191,18845
1,Old,0.6065,254423
2,Old Adult,0.4631,123210
3,Young Adult,0.5283,108728


- The very young group is too small to compared to others 
- The churn rate sometimes rises and sometimes falls there is no clear pattern 
- But Very young and Old show high churn rate compared to to others. 

In [22]:
query = """
SELECT Contract_Length, AVG(CASE WHEN Churn = 1 THEN 1 ELSE 0 END) AS churn_rate,COUNT(*) as Number_of_customer
FROM churn
GROUP BY Contract_Length;
"""
df = pd.read_sql(query,engine)
df

,Contract_Length,churn_rate,Number_of_customer
0,Annual,0.4609,198608
1,Monthly,0.9020,109234
2,Quarterly,0.4582,197364


#### Observations and Conclusion

1. The churn rate drops from 0.90 at the Monthly Contract_Length to 0.45 at the Quarterly Contract_Length	.

2. The percentage difference between Monthly and Quarterly Contract_Length is a High 64.7%.

3. The number of customers are comparable for each Contract_Length .

In [23]:
query = """
SELECT Subscription_Type, AVG(CASE WHEN Churn = 1 THEN 1 ELSE 0 END) AS churn_rate,COUNT(*) as Number_of_customer
FROM churn
GROUP BY Subscription_Type;
"""
df = pd.read_sql(query,engine)
df

,Subscription_Type,churn_rate,Number_of_customer
0,Basic,0.5689,164477
1,Premium,0.5475,170099
2,Standard,0.5497,170630


- There is not much difference in the churn rate for each class. 

In [24]:
query = """
SELECT Gender, AVG(CASE WHEN Churn = 1 THEN 1 ELSE 0 END) AS churn_rate,COUNT(*) as Number_of_customer
FROM churn
GROUP BY Gender;
"""
df = pd.read_sql(query,engine)
df

,Gender,churn_rate,Number_of_customer
0,Female,0.6489,224933
1,Male,0.4800,280273


- There is a slight difference in churn rate for female compared to male. 

## Final Feature Importance Summary

### Objective
To identify which customer behaviors are most strongly associated with churn, based on churn rate differences across feature segments.

### Evaluation Criteria
1. **Significant gap** in churn rates between segments
2. **Consistent pattern** (not random noise)
3. **Comparable segment sizes** (sufficient customer counts)

---

## Important Features (Strong Association with Churn)

| Rank | Feature | Key Finding | Churn Rate Gap | Customers (High-Risk Segment) |
|------|---------|-------------|----------------|-------------------------------|
| 1 | **Support Calls** | Churn jumps from ~0.29 (0-2 calls) to ~0.91 (5-10 calls) | ~62% | 183k |
| 2 | **Payment Delay** | Churn jumps from 0.47 (medium delay) to 0.94 (high delay) | ~47% | 111k |
| 3 | **Total Spend** | Churn drops sharply from 0.88 (low spend) to 0.41 (high spend) | ~47% | 151k |
| 4 | **Contract Length** | Monthly contracts have ~0.90 churn vs ~0.46 for annual/quarterly | ~44% | 109k |

---

## Less Important Features (Weak or Noisy Association)

| Feature | Churn Rate Range | Why Not Important |
|---------|------------------|--------------------|
| **Last Interaction (LI)** | 0.49 → 0.63 | Only ~25% difference |
| **Usage Frequency** | 0.53 → 0.60 | Small gap, unclear pattern |
| **Tenure** | 0.50 → 0.60 | Very narrow range (~10%) |
| **Subscription Type** | 0.55 – 0.57 | Almost no difference |
| **Age** | Varies widely | No consistent pattern |
| **Gender** | Female 0.65 / Male 0.48 | Moderate but weaker than top features |

---

## Final Conclusion

> **The most important predictors of churn are:**
> - High number of support calls
> - Long payment delays
> - Low total spend
> - Monthly contracts

These features show **large, consistent gaps** in churn rates across meaningful customer segments.

> Features like tenure, usage frequency, last interaction, subscription type, age, and gender have **weak or inconsistent associations** with churn and are less useful for predicting customer departure.

---

## Business Recommendations

| Feature | Actionable Insight |
|---------|---------------------|
| Support Calls (5+) | Proactively reach out to customers after 4+ support calls |
| Payment Delay (21+ days) | Flag accounts with >20 days delay for retention campaigns |
| Low Total Spend (<$508) | Investigate why low-spend customers leave; offer targeted incentives |
| Monthly Contracts | Encourage annual or quarterly plans with discounts or benefits |

---

**Next Steps:** Use these features to build a churn prediction model or design targeted retention interventions.